common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -496,6 +499,43 @@ public static void copyMemory(long srcAddr, byte[] dst, int dstIndex, long lengt
         PlatformDependent0.copyMemory(null, srcAddr, dst, BYTE_ARRAY_BASE_OFFSET + dstIndex, length);
     }
 
+    public static ByteBuffer allocateDirectNoCleaner(int capacity) {
+        if (DIRECT_MEMORY_COUNTER != null && DIRECT_MEMORY_COUNTER.addAndGet(capacity) > maxDirectMemory()) {
```

Comment:
Just checking for `DIRECT_MEMORY_COUNTER.addAndGet(capacity) > maxDirectMemory()` can cause over-allocation. (Difficult to trigger, but yea) Could be a CAS-like-loop for this.